### PR TITLE
fix(async queries): Remove "force" param on cached data retrieval

### PR DESCRIPTION
**Pull request by robdiciuccio**
### SUMMARY
When fetching cached data from the new `/api/v1/chart/data` endpoint in the async query lifecycle, the `force` param from the initial request was propagated to the cached data load, causing an error when force-reloading a chart's data. This PR removes the `force` param from the cached `QueryContext` data.

### TEST PLAN
Enable the `GLOBAL_ASYNC_QUERIES` feature flag and test force-reloading a chart using the new chart data API.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/common/query_context.py [hunk 1]
```diff
@@ -60,8 +60,8 @@ class QueryContext:
     queries: List[QueryObject]
     force: bool
     custom_cache_timeout: Optional[int]
-    result_type: utils.ChartDataResultType
-    result_format: utils.ChartDataResultFormat
+    result_type: Union[utils.ChartDataResultType, str]
+    result_format: Union[utils.ChartDataResultFormat, str]
```
Comment (villebro) [inline superset/common/query_context.py:64]: This `Union` shouldn't be necessary here. Are you getting linting errors without it?

File: superset/common/query_context.py [hunk 2]
```diff
@@ -71,23 +71,22 @@ def __init__(  # pylint: disable=too-many-arguments
         queries: List[Dict[str, Any]],
         force: bool = False,
         custom_cache_timeout: Optional[int] = None,
-        result_type: Optional[utils.ChartDataResultType] = None,
-        result_format: Optional[utils.ChartDataResultFormat] = None,
+        result_type: Optional[Union[utils.ChartDataResultType, str]] = None,
+        result_format: Optional[Union[utils.ChartDataResultFormat, str]] = None,
```
Comment (villebro) [inline superset/common/query_context.py:75]: Same here.

File: superset/common/query_context.py [hunk 3]
```diff
@@ -71,23 +71,22 @@ def __init__(  # pylint: disable=too-many-arguments
...
         self.datasource = ConnectorRegistry.get_datasource(
             str(datasource["type"]), int(datasource["id"]), db.session
         )
         self.queries = [QueryObject(**query_obj) for query_obj in queries]
         self.force = force
         self.custom_cache_timeout = custom_cache_timeout
-        self.result_type = result_type or utils.ChartDataResultType.FULL
-        self.result_format = result_format or utils.ChartDataResultFormat.JSON
+        self.result_type = result_type or utils.ChartDataResultType.FULL.value
+        self.result_format = result_format or utils.ChartDataResultFormat.JSON.value
```
Comment (villebro) [inline superset/common/query_context.py:84]: And here.

File: tests/query_context_tests.py [hunk 4]
```diff
@@ -73,13 +74,41 @@ def test_schema_deserialization(self):
...
+        query_context = ChartDataQueryContextSchema().load(payload)
+        query_object = query_context.queries[0]
+        query_cache_key = query_context.query_cache_key(query_object)
+
+        response = query_context.get_payload(cache_query_context=True)
+        cache_key = response["cache_key"]
+        assert cache_key
+
+        cached = cache_manager.cache.get(cache_key)
+        assert cached
```
Comment (villebro) [inline tests/query_context_tests.py:92]: nit: should we `assert cached is not None` here to highlight that we're not expecting a truthy value but something that isn't `None`?

**Review by villebro — COMMENTED** [review]
First pass comments, will test shortly

File: superset/common/query_context.py (hunk 1 above)
Comment (robdiciuccio) [reply superset/common/query_context.py:64]: Yes, marshmallow schema validation was failing with the default (unspecified) values for `result_type` and `result_format`. Essentially, marshmallow and mypy cannot agree here. I tried using https://pypi.org/project/marshmallow-enum/ here as well, to no avail.

**Comment by robdiciuccio** [comment]
Hm, yes @villebro. Technically, it is loading from cache on every async query. I'll see what can be done about the UI.

**Comment by robdiciuccio** [comment]
Ok, found the issue with legacy charts. The `/superset/explore_json/data/:cache_key` is setting `cache-control: public` 🤦 . I'll need to decouple the permission checks from the `@etag_cache` decorator.

File: superset/views/utils.py [hunk 5]
```diff
@@ -430,6 +431,23 @@ def is_owner(obj: Union[Dashboard, Slice], user: User) -> bool:
     return obj and user in obj.owners
 
 
+def check_resource_permissions(check_perms: Callable[..., Any],) -> Callable[..., Any]:
```
Comment (betodealmeida) [inline superset/views/utils.py:434]: Ah, nice... moving this out removes some of the complexity of the decorator.

**Review by betodealmeida — COMMENTED** [review]
@robdiciuccio can you explain why setting `cache-control: public` is causing problems? I thought the only difference here is that the response could be cached by proxies?

**Comment by robdiciuccio** [comment]
@betodealmeida the context here is force-reloading charts, which cannot serve cached data.

**Comment by villebro** [comment]
FYI @robdiciuccio flagging this for 1.0.1 (will retest shortly)

File: superset/common/query_context.py (hunk 1 above)
Comment (robdiciuccio) [reply superset/common/query_context.py]: I was able to get `marshmallow_enum` working, so the `Union` is no longer necessary.

**Review by villebro — APPROVED** [review]
Ok, I did further testing and turns out this was working correctly all along. The reason I thought there was a problem was because the `cached` pill always shows up, despite force reloading, as everything by definition comes from the cache (you did say that in your comment, my thinking was just too rooted in the old world). Well, I'll think about this some more, but in the meantime this is ready to merge IMO.

**Comment by robdiciuccio** [comment]
Thanks @villebro. Yeah, hiding the `cached` pill is a bit of a lift as all the async results are technically cached. The fact that the pill updates to "updated a few seconds ago" seems like an acceptable middleground for the time being.

**Pull request merged**